Video element with WebM Object URLs causes MediaError code 2

jyavenard opened the pull request:
#### af89b915c85561391bc98691eb511d1324b6985e
<pre>
Video element with WebM Object URLs causes MediaError code 2
<a href="https://bugs.webkit.org/show_bug.cgi?id=292618">https://bugs.webkit.org/show_bug.cgi?id=292618</a>
<a href="https://rdar.apple.com/151234095">rdar://151234095</a>

Reviewed by Youenn Fablet.

Don&apos;t throw an error if all content was loaded immediately as would happen
with a blob or data URL.

Additionally, don&apos;t cancel load if all data has already been received.

Test added.

* LayoutTests/media/media-vp8-webm-with-blob-expected.txt: Added.
* LayoutTests/media/media-vp8-webm-with-blob.html: Added.

Canonical link: <a href="https://commits.webkit.org/295802@main">https://commits.webkit.org/295802@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9d89ca8659f8d63e5a747e6ff3cc58731da7cabb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/106200 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/25949 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/16343 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/111397 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/56796 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/108239 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/26609 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/34452 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/80666 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/109204 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/21032 "Found 1 new test failure: media/video-unmuted-after-play-holds-sleep-assertion.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/95833 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/60992 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/20581 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/13936 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/56236 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/90391 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/13971 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/114258 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/33338 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/24584 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/89740 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/33702 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/92066 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/89441 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22801 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/34307 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/12122 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/28914 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/33263 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/38675 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/33009 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/36359 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/34607 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
```diff
@@ -1173,7 +1185,7 @@ - (void)removeAllVideoTargets;
     ALWAYS_LOG(LOGIDENTIFIER);
 
-    if (m_preload == MediaPlayer::Preload::MetaData)
+    if (m_preload == MediaPlayer::Preload::MetaData && !m_loadFinished)
```
I would guess that if load is finished, cancelLoad would be a no-op.
Can you detail why we need this in the commit message?

jyavenard replied:
> I would guess that if load is finished, cancelLoad would be a no-op. Can you detail why we need this in the commit message?

I did "Additionally, don't cancel load if all data has already been received."

youennf commented on LayoutTests/media/media-vp8-webm-with-blob.html:
```diff
@@ -0,0 +1,33 @@
+        (async () => {
+            const resp = await fetch('content/test-vp8.webm');
+            blob = await resp.blob();
+            run('video.src = URL.createObjectURL(blob)');
```
Can you also add a data URL test?

jyavenard replied:
Not sure it's adding much, code path taken would be the same.

youennf replied:
data URL loading is happening in process so timing can be a bit different.

jyavenard committed:
Video element with WebM Object URLs causes MediaError code 2
https://bugs.webkit.org/show_bug.cgi?id=292618
rdar://151234095

Reviewed by Youenn Fablet.

Don't throw an error if all content was loaded immediately as would happen
with a blob or data URL.

Additionally, don't cancel load if all data has already been received.

Test added.

* LayoutTests/media/media-vp8-webm-with-blob-expected.txt: Added.
* LayoutTests/media/media-vp8-webm-with-blob.html: Added.

Canonical link: https://commits.webkit.org/295802@main
